src/t_cose_mac_validate.c

```diff
@@ -0,0 +1,243 @@
+
+    /* The 0th tag is the only one that might identify the type of the
+     * CBOR we are trying to decode so it is handled special.
+     */
+    uTag = QCBORDecode_GetNthTagOfLast(decode_context, item_tag_index);
+    item_tag_index++;
+    if(me->option_flags & T_COSE_OPT_TAG_REQUIRED) {
+        /* The protocol that is using COSE says the input CBOR must
+         * be a COSE tag.
+         */
+        if(uTag != CBOR_TAG_COSE_SIGN1) {
+            return T_COSE_ERR_INCORRECTLY_TAGGED;
+        }
+    }
+    if(me->option_flags & T_COSE_OPT_TAG_PROHIBITED) {
+        /* The protocol that is using COSE says the input CBOR must
+         * not be a COSE tag.
+         */
+        if(uTag == CBOR_TAG_COSE_SIGN1) {
+            return T_COSE_ERR_INCORRECTLY_TAGGED;
+        }
+    }
+    /* If the protocol using COSE doesn't say one way or another about the
+     * tag, then either is OK.
+     */
+
+
+    /* Initialize auTags, the returned tags, to CBOR_TAG_INVALID64 */
+#if CBOR_TAG_INVALID64 != 0xffffffffffffffff
+#error Initializing return tags array
+#endif
+    memset(me->auTags, 0xff, sizeof(me->auTags));
+
+    returned_tag_index = 0;
+
+    if(uTag != CBOR_TAG_COSE_SIGN1) {
+        /* Never return the tag that this code is about to process. Note
+         * that you can sign a COSE_SIGN1 recursively. This only takes out
+         * the one tag layer that is processed here.
+         */
+        me->auTags[returned_tag_index] = uTag;
+        returned_tag_index++;
+    }
+
+    while(1) {
+        uTag = QCBORDecode_GetNthTagOfLast(decode_context, item_tag_index);
+        item_tag_index++;
+        if(uTag == CBOR_TAG_INVALID64) {
+            break;
+        }
+        if(returned_tag_index > T_COSE_MAX_TAGS_TO_RETURN) {
+            return T_COSE_ERR_TOO_MANY_TAGS;
+        }
+        me->auTags[returned_tag_index] = uTag;
+        returned_tag_index++;
+    }
+
+    return T_COSE_SUCCESS;
+}
+
+/**
+ * \file t_cose_mac_validate.c
+ *
+ * \brief This verifies t_cose Mac authentication structure without a recipient
+ *        structure.
+ *        Only HMAC is supported so far.
+ */
+
+/*
+ * Public function. See t_cose_mac.h
+ */
+enum t_cose_err_t t_cose_mac_validate_private(struct t_cose_mac_validate_ctx *context,
+                                            struct q_useful_buf_c         cose_mac,
+                                            struct q_useful_buf_c         aad,
+                                            bool                          payload_is_detached,
+                                            struct q_useful_buf_c        *payload,
+                                            struct t_cose_header_param  **return_params)
+{
+    QCBORDecodeContext            decode_context;
+    struct q_useful_buf_c         protected_parameters;
+    QCBORError                    qcbor_error;
+    struct t_cose_header_param    params_arr[4];
+    struct header_param_storage   params = {0,params_arr};
```

Comment:
Should be {4, params_arr}.  4 is the size of the array that can be used.  4 should be T_COSE_NUM_VERIFY_DECODE_HEADERS
params must be part of t_cose_mac_validate_ctx so you can return it to the caller.